### PR TITLE
implementation of rmsprop that uses epsilon values to help stabilise

### PR DESCRIPTION
**Pull request by willwilliams**
- implementation is reasonably efficient
- could be that clone is unnecessary and the square gradients can be calculated in place
- no NAG or momentum (because I've never found it to be helpful)
- eplison values included (because I've always found them to be very useful and the algorithm appears to be sensitive to setting them correctly)

### CODE REVIEW

**Comment by jonathantompson** [comment]
Actually, is the implementation correct?

It's supposed to be a moving average filter so the update should be:

m = alpha * m + (1 - alpha) * (df/dx)^2

The epsilon is OK.  I found to get RMSPROP to actually working you need to
have a gain_max and gain_min term.  That is you need to do a elementwise
clamp of the one_over_rms vector before you apply it as the learning rate
multiplier.  This is really important for cases where you get stuck in a
saddle point and the derivative along some direction goes close to zero...
 1/tiny will result in numerical instability.

**Comment by willwilliams** [comment]
So as far as I can see my implementation does exactly what you describe. Please correct me if that's wrong!

I've added the clamping but with fairly arbitrary default values. It seems like there's lots of knowledge in the community about how to get this working well in practice so it would be good if the implementation could make use of the best default values for a wide range of problems.

**Comment by jonathantompson** [comment]
Oh OK.  Sorry, I didn't see line 40 in the original commit :-(  My bad.

Thanks for adding the clamping...  I find sometimes it's not needed at all, but sometimes it's essential, so it's great that it's now in there.

One minor comment.  I think instead of dfdx:clone(), maybe add a dfdx_sq variable to the optimState and do a resize and copy instead.  Even if we need to have the extra memory, it would be nice to at least avoid the clone (so that we're not having to rely on the garbage collector to clean it up).

Thanks for the commit!

File: rmsprop.lua [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    state.dfdx_sq = state.dfdx_sq or torch.Tensor():typeAs(dfdx):resizeAs(dfdx)
+
+    -- (3) calculate new mean squared values
+    torch.cmul(state.dfdx_sq, dfdx, dfdx)
+    state.m:mul(alpha)
+    state.m:add(state.dfdx_sq:mul(1.0-alpha))
+    state.m:add(epsilon2)
+
+    -- (4) perform update
+    local one_over_rms = torch.pow(state.m, -0.5)
```
Comment (soumith) [inline rmsprop.lua]: repeated malloc, that can be avoided with a temporary buffer (that is stored in state)

File: rmsprop.lua [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    one_over_rms:clamp(min_gain, max_gain)
+    local update = torch.cmul(torch.mul(dfdx,lr), one_over_rms)
```
Comment (soumith) [inline rmsprop.lua]: same ^^

**Comment by soumith** [comment]
lgtm except for a couple of comments.

**Comment by soumith** [comment]
Thanks Will. @koraykv let me know if you find anything incorrect, looks okay to me.Thanks Will. @koraykv let me know if you find anything incorrect, looks okay to me.

**Pull request merged**